File: packages/core/src/cleaner.ts
import { mkdirSync, readFileSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import {
  AnalysisReport,
  analyzeText,
  compareReports,
} from "./analyzer";
import { DEFAULT_PROFILE, Profile, getProfile, profileToDict } from "./config";

const LIST_PATTERN = /^(\s*[-*]|\s*\d+\.\s+)/;
const URL_PATTERN = /^https?:\/\/[^\s]+/;

export function normalizeText(text: string, profile?: Profile): string {
  const emDash = profile?.normalize_em_dash ?? true;
  const spacePunct = profile?.normalize_space_before_punctuation ?? true;
  let t = text.replace(/\x0c|\f/g, "\n");
  if (spacePunct) {
    t = t.replace(/[ \t]+,[ \t]*/g, ", ");
    t = t.replace(/[ \t]+\.[ \t]*/g, ". ");
  }
  if (emDash) {
    t = t.replace(/[ \t]*—[ \t]*|[ \t]*--[ \t]*/g, " — ");
  }
  return t;
}

function isList(line: string): boolean {
  return LIST_PATTERN.test(line);
}

function isUrl(line: string): boolean {
  return URL_PATTERN.test(line);
}

function isHeading(line: string, profile: Profile): boolean {
  const s = line.trim();
  return s.startsWith("#") || new RegExp(profile.chapter_pattern, "i").test(s);
}

function isSectionHeading(line: string, profile: Profile): boolean {
  const s = line.trim();
  if (new RegExp(profile.section_pattern).test(s)) return true;
  if (profile.subsection_pattern && new RegExp(profile.subsection_pattern).test(s)) return true;
  return false;
}

function isTableRow(line: string): boolean {
  const s = line.trim();
  if ((s.match(/\|/g) ?? []).length >= 2) return true;
  return (s.match(/\d+[.,]?\d*/g) ?? []).length >= 3 && s.length < 60;
}

function shouldJoinUrl(previous: string, current: string, profile: Profile): boolean {
  if (!previous) return false;
  if (previous.startsWith("http") && previous.endsWith("-"))
    return !!current && !isHeading(current, profile);
  if (previous.startsWith("http"))
    return current.includes("/") || current.includes(".") || /^(en|id)\//.test(current);
  return false;
}

function standardizeHeading(line: string, profile: Profile): string {
  if (isHeading(line, profile) && !line.startsWith("#")) return `# ${line}`;
  if (isSectionHeading(line, profile) && !line.startsWith("#")) return `## ${line}`;
  return line;
}

export function mergeAndStructure(text: string, profile?: Profile): string {
  const p = profile ?? getProfile(DEFAULT_PROFILE);
  const rawLines = normalizeText(text, p).split("\n");
  const processed: string[] = [];
  const paragraph: string[] = [];

  const flushParagraph = () => {
    if (paragraph.length) {
      processed.push(paragraph.join(" "));
      paragraph.length = 0;
    }
  };

  const endings = p.sentence_endings as string[];

  for (const rawLine of rawLines) {
    const line = rawLine.trim().replace(/\s+/g, " ");

    if (!line) {
      if (paragraph.length && endings.some(e => paragraph[paragraph.length - 1].endsWith(e))) {
        flushParagraph();
        processed.push("");
      }
      continue;
    }

    if (paragraph.length && shouldJoinUrl(paragraph[paragraph.length - 1], line, p)) {
      let prev = paragraph.pop()!;
      if (prev.endsWith("-")) prev = prev.slice(0, -1);
      paragraph.push(prev + line);
      continue;
    }

    const structural =
      isHeading(line, p) || isSectionHeading(line, p) || isList(line) || isUrl(line) || isTableRow(line);

    if (!structural && processed.length && shouldJoinUrl(processed[processed.length - 1], line, p)) {
      let prev = processed.pop()!;
      if (prev.endsWith("-")) prev = prev.slice(0, -1);
      processed.push(prev + line);
      continue;
    }

    if (structural) {
      flushParagraph();
      const out = standardizeHeading(line, p);
      if (isTableRow(out)) {
        processed.push(out);
      } else {
        const prevIsList = processed.length > 0 && isList(processed[processed.length - 1]);
        if (processed.length && processed[processed.length - 1] !== "" && !(isList(out) && prevIsList)) {
          processed.push("");
        }
        processed.push(out);
        if (!isList(out) && !isUrl(out)) processed.push("");
      }
      continue;
    }

    paragraph.push(line);
  }

  flushParagraph();
  return processed.join("\n").replace(/\n{3,}/g, "\n\n").trim() + "\n";
}

export type Manifest = {
  input_path: string;
  output_path: string;
  dry_run: boolean;
  timestamp: string;
  profile: ReturnType<typeof profileToDict>;
  before: AnalysisReport;
  after: AnalysisReport;
  comparison: ReturnType<typeof compareReports>;
  warnings: string[];
  errors: string[];
  severity: "ok" | "warning" | "error";
};

export function cleanManuscript(
  inputPath: string,
  outputPath: string,
  profile?: Profile,
  strict = false,
  dryRun = false
): { cleanText: string; manifest: Manifest; manifestPath: string } {
  const p = profile ?? getProfile(DEFAULT_PROFILE);
  const rawText = readFileSync(inputPath, "utf-8");
  const before = analyzeText(rawText, p, false);
  const cleanText = mergeAndStructure(rawText, p);
  const after = analyzeText(cleanText, p, strict);
  const comparison = compareReports(before, after, p, strict);

  if (!dryRun) {
    mkdirSync(dirname(outputPath), { recursive: true });
    writeFileSync(outputPath, cleanText, "utf-8");
  }

  const allWarnings = [...after.warnings, ...comparison.warnings];
  const allErrors = [...after.errors, ...comparison.errors];
  const severity: "ok" | "warning" | "error" =
    after.severity === "error" || comparison.severity === "error"
      ? "error"
      : after.severity === "warning" || comparison.severity === "warning"
      ? "warning"
      : "ok";

  const manifest: Manifest = {
    input_path: inputPath,
    output_path: outputPath,
    dry_run: dryRun,
    timestamp: new Date().toISOString(),
    profile: profileToDict(p),
    before,
    after,
    comparison,
    warnings: allWarnings,
    errors: allErrors,
    severity,
  };

  const manifestPath = join(dirname(outputPath), "manifest.json");
  return { cleanText, manifest, manifestPath };
}

File: manuscript/cleaner.py
import re
from datetime import datetime, timezone
from pathlib import Path

from .analyzer import analyze_text, compare_reports, is_heading, is_section_heading, is_table_row
from .config import get_profile


LIST_PATTERN = re.compile(r"^(\s*[-*]|\s*\d+\.\s+)")
URL_PATTERN = re.compile(r"^https?://[^\s]+")


def normalize_text(text, profile=None):
    em_dash = profile.normalize_em_dash if profile is not None else True
    space_punct = profile.normalize_space_before_punctuation if profile is not None else True
    text = text.replace("\x0c", "\n").replace("\f", "\n")
    if space_punct:
        text = re.sub(r"[ \t]+,[ \t]*", ", ", text)
        text = re.sub(r"[ \t]+\.[ \t]*", ". ", text)
    if em_dash:
        text = re.sub(r"[ \t]*—[ \t]*|[ \t]*--[ \t]*", " — ", text)
    return text


def is_list(line):
    return bool(LIST_PATTERN.match(line))


def is_url(line):
    return bool(URL_PATTERN.match(line))


def should_join_url(previous, current):
    if not previous:
        return False
    if previous.startswith("http") and previous.endswith("-"):
        return bool(current) and not is_heading(current, get_profile())
    if previous.startswith("http"):
        return "/" in current or "." in current or current.startswith(("en/", "id/"))
    return False


def standardize_heading(line, profile):
    if is_heading(line, profile) and not line.startswith("#"):
        return f"# {line}"
    if is_section_heading(line, profile) and not line.startswith("#"):
        return f"## {line}"
    return line


def flush_paragraph(processed_lines, current_paragraph):
    if current_paragraph:
        processed_lines.append(" ".join(current_paragraph))
        current_paragraph.clear()


def merge_and_structure(text, profile=None):
    profile = profile or get_profile()
    raw_lines = normalize_text(text, profile=profile).splitlines()
    processed_lines = []
    current_paragraph = []

    for raw_line in raw_lines:
        line = re.sub(r"\s+", " ", raw_line.strip())

        if not line:
            if current_paragraph and current_paragraph[-1].endswith(profile.sentence_endings):
                flush_paragraph(processed_lines, current_paragraph)
                processed_lines.append("")
            continue

        if current_paragraph and should_join_url(current_paragraph[-1], line):
            previous = current_paragraph.pop()
            if previous.endswith("-"):
                previous = previous[:-1]
            current_paragraph.append(previous + line)
            continue

        structural = (
            is_heading(line, profile)
            or is_section_heading(line, profile)
            or is_list(line)
            or is_url(line)
            or is_table_row(line)
        )

        if not structural and processed_lines and should_join_url(processed_lines[-1], line):
            previous = processed_lines.pop()
            if previous.endswith("-"):
                previous = previous[:-1]
            processed_lines.append(previous + line)
            continue

        if structural:
            flush_paragraph(processed_lines, current_paragraph)
            line = standardize_heading(line, profile)
            if is_table_row(line):
                processed_lines.append(line)
            else:
                previous_is_list = bool(processed_lines and is_list(processed_lines[-1]))
                if processed_lines and processed_lines[-1] != "" and not (
                    is_list(line) and previous_is_list
                ):
                    processed_lines.append("")
                processed_lines.append(line)
                if not is_list(line) and not is_url(line):
                    processed_lines.append("")
            continue

        if current_paragraph:
            current_paragraph.append(line)
        else:
            current_paragraph.append(line)

    flush_paragraph(processed_lines, current_paragraph)
    final_text = "\n".join(processed_lines)
    final_text = re.sub(r"\n{3,}", "\n\n", final_text).strip() + "\n"
    return final_text


def clean_manuscript(input_path, output_path, profile=None, strict=False, dry_run=False):
    profile = profile or get_profile()
    input_path = Path(input_path)
    output_path = Path(output_path)
    raw_text = input_path.read_text(encoding="utf-8")
    before = analyze_text(raw_text, profile=profile, strict=False)
    clean_text = merge_and_structure(raw_text, profile=profile)
    after = analyze_text(clean_text, profile=profile, strict=strict)
    comparison = compare_reports(before, after, profile=profile, strict=strict)
    if not dry_run:
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(clean_text, encoding="utf-8")
    manifest = {
        "input_path": str(input_path),
        "output_path": str(output_path),
        "dry_run": dry_run,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "profile": profile.to_dict(),
        "before": before,
        "after": after,
        "comparison": comparison,
        "warnings": after["warnings"] + comparison["warnings"],
        "errors": after["errors"] + comparison["errors"],
        "severity": "error"
        if after["severity"] == "error" or comparison["severity"] == "error"
        else "warning"
        if after["severity"] == "warning" or comparison["severity"] == "warning"
        else "ok",
    }
    manifest_path = output_path.parent / "manifest.json"
    return clean_text, manifest, manifest_path
